File: Backtracking/leetcode_17.py
# Problem Link ====>>https://leetcode.com/problems/letter-combinations-of-a-phone-number/description/

def letterCombinations(digits):
    if not digits:
        return []

    mappings = {
        "2": "abc",
        "3": "def",
        "4": "ghi",
        "5": "jkl",
        "6": "mno",
        "7": "pqrs",
        "8": "tuv",
        "9": "wxyz"
    }

    result = []
    current = []
    backtrack(mappings, digits, 0, current, result)
    return result


def backtrack(mappings, digits, index, current, result):
    if index == len(digits):
        result.append("".join(current))
        return

    letters = mappings[digits[index]]
    for letter in letters:
        current.append(letter)
        backtrack(mappings, digits, index + 1, current, result)
        current.pop()


digits = "23"
combinations = letterCombinations(digits)
print("Letter combinations of phone number", digits, ":")
for combination in combinations:
    print(combination)

File: Backtracking/leetcode_17.cpp

// Problem Link ====>>https://leetcode.com/problems/letter-combinations-of-a-phone-number/description/

#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution
{
public:
    vector<string> letterCombinations(string digits)
    {
        vector<string> result;

        if (digits.empty())
        {
            return result;
        }

        unordered_map<char, string> mappings = {
            {'2', "abc"},
            {'3', "def"},
            {'4', "ghi"},
            {'5', "jkl"},
            {'6', "mno"},
            {'7', "pqrs"},
            {'8', "tuv"},
            {'9', "wxyz"},
        };

        string current;
        backtrack(mappings, digits, 0, current, result);
        return result;
    }

private:
    void backtrack(const unordered_map<char, string> &mappings, const string &digits, int index, string &current, vector<string> &result)
    {
        if (index == digits.length())
        {
            result.push_back(current);
            return;
        }

        const string &letters = mappings.at(digits[index]);
        for (char letter : letters)
        {
            current.push_back(letter);
            backtrack(mappings, digits, index + 1, current, result);
            current.pop_back();
        }
    }
};

int main()
{
    string digits = "23";
    Solution solution;
    vector<string> combinations = solution.letterCombinations(digits);

    cout << "Letter combinations of phone number " << digits << ":" << endl;
    for (string combination : combinations)
    {
        cout << combination << endl;
    }

    return 0;
}
